File: abeln-hb-lattice-2e567b22ecd1/source/MolBox.cpp
#include "MolBox.hpp"

#include "Chain.hpp"
#include "Lattice.hpp"
#include "Native.hpp"
#include "Stats.hpp"
#include "App.hpp"
#include "Moves.hpp"

MolBox::MolBox(Lattice * mainL){
  //beta = 0.08;
  currentConfig =0;
  numConfigs =1000;
  box = new Lattice;
  mainLattice = mainL;
  for(int i =0;i<MAX_CONFIGS;i++){
    configSet[i]=NULL;
  }
}

void MolBox::setAA(string fn, int indx){
  box->setAA(fn,indx);
}

/*void MolBox::setBeta(double b){
  beta=b;
  }*/

double MolBox::getBeta(){
  return box->getBetaMoves();
}


void MolBox::setMolecule(Chain * chain){
  // lattice should be clean 

  molecule = new Chain(box,0);
  box->chains[0]=molecule;
  molecule->frozen = false;
  molecule->locked = false;
  molecule->N = chain->N;
  
  // copy and insert each residue
  for(int n=0;n<chain->N;n++){
    Residue * res = new Residue;
    res->pos.x = chain->residues[n]->pos.x;
    res->pos.y =chain->residues[n]->pos.y;
    res->pos.z =chain->residues[n]->pos.z;
    res->spin = chain->residues[n]->spin;
    res->state = chain->residues[n]->state;
    res->aa =chain->residues[n]->aa;
    sequence[n]=res->aa;
    res->n =chain->residues[n]->n;
    res->chainNum=0;  
    molecule->residues[n]=res;
    box->r[res->pos.x][res->pos.y][res->pos.z] = molecule->residues[n];
  }
  molecule->setFwdBkwdSpins();
  box->nChains = 1;
  if(App::nativeStats){
    box->native = new Native(box);
  }
  box->stats.getLatticeStats(box);
  box->energyMap = NULL;// should not record eMap
}



void MolBox::createNewSet(double b,int num_configs, bool betaChanged){
  // need to set Stats::l
  //cout<< "CREATING NEW SET "<<endl;
  box->setBetaMoves(b);
  //molecule->setBeta(beta);

  if(betaChanged){   
    //cout<<"EQUIL BETA CHANGE"<<endl;
    equilibrate(EQUIL_STEPS);
  }
 
  currentConfig=0;
  if(num_configs > MAX_CONFIGS)num_configs = MAX_CONFIGS;
  numConfigs=num_configs;
 
  for(int nc =0;nc<numConfigs;nc++){
    //cout<< nc <<endl;
    equilibrate((int) floor(100 * drand48()));
    recordConfig(nc);         
  } 
}

void  MolBox::printConfig(int c){
  Config * config = configSet[c];
  cout<<"configuration"<< c<<endl;
  for(int n=0;n<config->nTotal;n++){
    cout<<"config res:"<<n<<" "<<config->positions[n].x<<" "<<config->positions[n].y<<" "<<config->positions[n].z<<endl;
  }
}



Config * MolBox::getNewConfig(){
  currentConfig++;
  if(currentConfig >= numConfigs){
    createNewSet(box->getBetaMoves(),numConfigs, false);
  }
  // perhaps better to choose @random?
  int c = (int) floor((numConfigs-1) * drand48());
  //cout<<"config no: "<<c<<endl;
  //printConfig(c);
  return configSet[c ];  
}
 

void MolBox::equilibrate(int nsteps){
  //cout<<"EQUILIBIRATING"<<endl;
  for(int i =0;i<nsteps;i++){
    //molecule->localMove();
    Moves::localMove(box,molecule);
    if(drand48() < 0.05 ){
      Moves::rotatePoint(box,molecule);
    }
    if(drand48() < 0.5){
      Moves::changeStrandCoil(box,molecule);
    }else{
      Moves::shuffleSpin(box,molecule);
    }
  }
}


void MolBox::recordConfig(int insertAt){
  //cout<<"recording config"<<endl;
  if(configSet[insertAt] !=NULL){
    delete configSet[insertAt];
    configSet[insertAt] = NULL;
  }
  
  Config * cc = new Config;
 
  for(int n=0; n<molecule->N;n++){
    // perhaps needs normalising ?

    cc->positions[n] = molecule->residues[n]->pos;
    cc->spins[n] =  molecule->residues[n]->spin;
    cc->states[n] = molecule->residues[n]->state;
    cc->nTotal = molecule->N;
  }

  configSet[insertAt]= cc;
}

File: nelman-kmcLATTICEMODEL/source/MolBox.py
from typing import List, Optional
import random
from Lattice import Lattice
from Chain import Chain
from Config import Config
from Moves import localMove, rotatePoint, changeStrandCoil, shuffleSpin

MAX_CONFIGS = 1000
EQUIL_STEPS = 1000

class MolBox:
    def __init__(self, main_l: Lattice):
        self.current_config = 0
        self.num_configs = 1000
        self.box = Lattice()
        self.main_lattice = main_l
        self.config_set = [None] * MAX_CONFIGS
        self.sequence = [0] * MAX_RES
        self.molecule = None
    
    def setAA(self, fn: str, indx: int):
        self.box.setAA(fn, indx)
    
    def getBeta(self) -> float:
        return self.box.getBetaMoves()
    
    def setMolecule(self, chain: Chain):
        self.molecule = Chain(self.box, 0)
        self.box.chains[0] = self.molecule
        self.molecule.frozen = False
        self.molecule.locked = False
        self.molecule.N = chain.N
        
        for n in range(chain.N):
            res = Residue()
            res.pos = chain.residues[n].pos.copy()
            res.spin = chain.residues[n].spin.copy()
            res.state = chain.residues[n].state
            res.aa = chain.residues[n].aa
            self.sequence[n] = res.aa
            res.n = chain.residues[n].n
            res.chainNum = 0
            self.molecule.residues[n] = res
            self.box.r[res.pos.x][res.pos.y][res.pos.z] = res
        
        self.molecule.setFwdBkwdSpins()
        self.box.nChains = 1
        self.box.energyMap = None
    
    def createNewSet(self, b: float, num_configs: int, beta_changed: bool):
        self.box.setBetaMoves(b)
        
        if beta_changed:
            self.equilibrate(EQUIL_STEPS)
        
        self.current_config = 0
        self.num_configs = min(num_configs, MAX_CONFIGS)
        
        for nc in range(self.num_configs):
            self.equilibrate(random.randint(0, 99))
            self.recordConfig(nc)
    
    def getNewConfig(self) -> Optional[Config]:
        self.current_config += 1
        if self.current_config >= self.num_configs:
            self.createNewSet(self.box.getBetaMoves(), self.num_configs, False)
        return self.config_set[random.randint(0, self.num_configs - 1)]
    
    def equilibrate(self, n_steps: int):
        for _ in range(n_steps):
            localMove(self.box, self.molecule)
            if random.random() < 0.05:
                rotatePoint(self.box, self.molecule)
            if random.random() < 0.5:
                changeStrandCoil(self.box, self.molecule)
            else:
                shuffleSpin(self.box, self.molecule)
    
    def recordConfig(self, insert_at: int):
        if self.config_set[insert_at] is not None:
            del self.config_set[insert_at]
        
        cc = Config()
        for n in range(self.molecule.N):
            cc.positions[n] = self.molecule.residues[n].pos.copy()
            cc.spins[n] = self.molecule.residues[n].spin.copy()
            cc.states[n] = self.molecule.residues[n].state
            cc.nTotal = self.molecule.N
        
        self.config_set[insert_at] = cc
